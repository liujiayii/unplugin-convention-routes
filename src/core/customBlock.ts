import type { SFCBlock, SFCDescriptor } from '@vue/compiler-sfc'
import type { CustomBlock, ParsedJSX, ResolvedOptions } from './types'
import fs from 'node:fs'

import JSON5 from 'json5'

import { importModule } from 'local-pkg'
import { parse as YAMLParser } from 'yaml'
// @ts-expect-error no type
import extractComments from 'extract-comments'
import { debug } from './utils'

const routeJSXReg = /^\s+(route)\s+/gm

export function parseJSX(code: string): ParsedJSX[] {
  return extractComments(code).slice(0, 1).filter((comment: ParsedJSX) => routeJSXReg.test(comment.value) && comment.value.includes(':') && comment.loc.start.line === 1)
}

export function parseYamlComment(code: ParsedJSX[], path: string): CustomBlock {
  return code.reduce((memo, item) => {
    const { value } = item
    const v = value.replace(routeJSXReg, '')
    debug.routeBlock(`use ${v} parser`)
    try {
      const yamlResult = YAMLParser(v)

      return {
        ...memo,
        ...yamlResult,
      }
    }
    catch (err: any) {
      throw new Error(`Invalid YAML format of comment in ${path}\n${err.message}`)
    }
  }, {})
}

export async function parseSFC(code: string): Promise<SFCDescriptor> {
  try {
    const { parse } = await importModule('@vue/compiler-sfc') as typeof import('@vue/compiler-sfc')
    return parse(code, {
      pad: 'space',
    }).descriptor
    // for @vue/compiler-sfc ^2.7
    || (parse as any)({
      source: code,
    })
  }
  catch {
    throw new Error('[unplugin-pages] Vue3\'s "@vue/compiler-sfc" is required.')
  }
}

export function parseCustomBlock(block: SFCBlock, filePath: string, options: ResolvedOptions): any {
  const lang = block.lang ?? options.routeBlockLang

  debug.routeBlock(`use ${lang} parser`)

  if (lang === 'json5') {
    try {
      return JSON5.parse(block.content)
    }
    catch (err: any) {
      throw new Error(`Invalid JSON5 format of <${block.type}> content in ${filePath}\n${err.message}`)
    }
  }
  else if (lang === 'json') {
    try {
      return JSON.parse(block.content)
    }
    catch (err: any) {
      throw new Error(`Invalid JSON format of <${block.type}> content in ${filePath}\n${err.message}`)
    }
  }
  else if (lang === 'yaml' || lang === 'yml') {
    try {
      return YAMLParser(block.content)
    }
    catch (err: any) {
      throw new Error(`Invalid YAML format of <${block.type}> content in ${filePath}\n${err.message}`)
    }
  }
}

export async function getRouteBlock(path: string, options: ResolvedOptions) {
  const content = fs.readFileSync(path, 'utf8')

  const parsedSFC = await parseSFC(content)
  const blockStr = parsedSFC?.customBlocks.find(b => b.type === 'route')

  const parsedJSX = parseJSX(content)

  if (!blockStr && parsedJSX.length === 0)
    return

  let result

  if (blockStr)
    result = parseCustomBlock(blockStr, path, options) as CustomBlock

  if (parsedJSX.length > 0)
    result = parseYamlComment(parsedJSX, path) as CustomBlock

  return result
}
