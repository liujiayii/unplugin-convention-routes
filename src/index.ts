import type { UnpluginFactory } from 'unplugin'
import type { Options, UserOptions } from './core/types'

import path from 'node:path'

import process from 'node:process'
import { createUnplugin } from 'unplugin'
import { PageContext } from './core/context'

export const unpluginFactory: UnpluginFactory<Options | undefined> = (userOptions: UserOptions = {}) => {
  const ctx: PageContext = new PageContext(userOptions)
  //  ctx.setLogger(config.logger)
  return ({
    name: 'unplugin-pages',
    buildEnd() {
      ctx.watcher?.close()
    },
    resolveId(id) {
      //  console.log('resolveId', id)
      if (['~unplugin-pages/react', '~unplugin-pages/vue', '~unplugin-pages/solid'].includes(id)) {
        return path.resolve(process.cwd(), 'node_modules', '~unplugin-pages', `${id.split('/')[1]}.js`)
      }
      return null
    },
  })
}

export const unplugin = /* #__PURE__ */ createUnplugin(unpluginFactory)

export default unplugin
