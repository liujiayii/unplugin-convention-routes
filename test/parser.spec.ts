import { resolve } from 'node:path'
import { getRouteBlock } from '../src/customBlock'
import { resolveOptions } from '../src/options'

const options = resolveOptions({})

describe('parser', () => {
  it('custom block', async () => {
    const path = resolve('./examples/vue/src/pages/blog/[id].vue')
    const routeBlock = await getRouteBlock(path, options)
    expect(routeBlock).toMatchSnapshot()
  })

  it('jsx yaml comment', async () => {
    const path = resolve('./examples/vue/src/pages/jsx.jsx')
    const routeBlock = await getRouteBlock(path, options)
    expect(routeBlock).toMatchSnapshot()
  })
})
