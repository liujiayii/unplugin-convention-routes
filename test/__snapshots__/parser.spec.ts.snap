// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parser > custom block 1`] = `
{
  "meta": {
    "requiresAuth": false,
  },
  "name": "blog-id",
}
`;

exports[`parser > jsx yaml comment 1`] = `
{
  "meta": {
    "id": 1234,
    "requiresAuth": false,
  },
  "name": "blog-id",
}
`;
